### PR TITLE
381 add ability to split native vs wrapped carbon orders

### PR DESCRIPTION
**Pull request by Lesigh-3100**
**Splitting Carbon Trades**

New class:
CarbonTradeSplitter

Current problem: 

In the main arb processing flow in bot.py _handle_trade_instructions, all trades that include Carbon deployments have been aggregated into a single trade action, which will cause trades to fail in cases in which a single trade action includes strategies that contain a mix of wrapped/native tokens (paired with the same token). a mix of strategies being traded on include native gas token, and some include wrapped gas token, or the trades occur on different carbon deployments. 

**Example 1:**
- Strategy 1 is trading ETH > TKN on Carbon
- Strategy 2 is trading WETH > TKN  on Carbon
- Trade TKN > WETH on Uniswap

_Current output:_
Trade 1: Carbon, WETH > TKN [Strategy 1, 2]
Trade 2: Uniswap, TKN > WETH

_Expected output:_
Trade 1: Carbon, ETH > TKN [Strategy 1]
Trade 2: Carbon, WETH > TKN [Strategy 2]
Trade 3: Uniswap, TKN > WETH

**Example 2:**
- Strategy 1 is trading ETH > TKN on Carbon
- Strategy 2 is trading ETH > TKN on Carbon
- Strategy 3 is trading WETH > TKN on Carbon
- Strategy 4 is trading ETH > TKN  on Graphene
- - Trade TKN > WETH on Uniswap

_Current output:_
Trade 1: Carbon, ETH > TKN [Strategy 1, 2, 3, 4]
Trade 2: Uniswap, TKN > WETH

_Expected output:_
Trade 1: Carbon, ETH > TKN [Strategy 1, 2]
Trade 2: Carbon, WETH > TKN [Strategy 3]
Trade 3: Graphene, ETH > TKN [Strategy 4]
Trade 4: Uniswap, TKN > WETH

**Updated Wrap/Unwrap Algorithm**

New class:
WrapUnwrapProcessor

To support the split trade upgrade, the wrap/unwrap algorithm has been updated to be significantly more robust. It can now handle much more complex routes (such as when you split Carbon trades!).

### CODE REVIEW

File: fastlane_bot/helpers/routehandler.py [hunk 1]
```diff
@@ -440,6 +438,103 @@ def add_wrap_or_unwrap_trades_to_route(self, trade_instructions: List[TradeInstr
                                                                              source_amount=0)))
         return new_route_struct
 
+    def add_wrap_or_unwrap_trades_to_route_v4(self, trade_instructions: List[TradeInstruction],
```
Comment (mikewcasale) [inline fastlane_bot/helpers/routehandler.py:441]: ### Review of `add_wrap_or_unwrap_trades_to_route_v4` 

- **`Readability`**
    - The function is lengthy and does a lot within a single block of code, which hampers readability. It mixes different levels of abstraction, from iterating over input structures to manipulating balances and creating new route structures.
    - Variable names like `__tkn_in` and `__tkn_out` in `add_subtract_wrapped_native` are not immediately clear. Using more descriptive names would improve readability.
    - Inline comments are used, but the function's overall complexity still makes it hard to follow.

- **`Maintainability`**
    - The monolithic structure makes the function hard to maintain. Changes in the logic related to any step (e.g., balance tracking, route segmentation) require careful navigation and modification of the function, increasing the risk of introducing errors.
    - Hard-coded checks and assertions within the function can make it brittle in response to changes in business logic or data structure.
    - Sub-functions like `add_subtract_wrapped_native` are unreachable from the main function, making it harder to test and maintain.

- **`DRY Principle`**
    - The function contains multiple repeated operations, which should be extracted into separate methods or functions to adhere to the Don't Repeat Yourself (DRY) principle.

- **`Single Responsibility Principle`**
    - The function could benefit from following the Single Responsibility Principle more closely. Consider breaking it down into smaller functions or methods, each handling a specific part of the logic.

- **`Styling`**
    - The function should adhear to PEP8 styling guidelines, and use Google-style docstrings for documentation.

    
#### Suggestions for Improvement

- Break down `add_wrap_or_unwrap_trades_to_route_v4` into multiple methods within a class to improve the readability significantly.
- Each method should follow the single responsibility principle, making the code easier to understand.
- Method names and parameters are descriptive, aiding in understanding the code's purpose at a glance.
- A class structure would make it easier to extend functionality, such as adding new methods for handling different types of trades or adjusting the logic for wrap/unwrap operations.
- Add unit tests for each method in the class to improve the code's maintainability. 

### Example Improvement

For instance, refactoring the `add_wrap_or_unwrap_trades_to_route_v4` function into a class `WrapUnwrapProcessor` could look like this (Note: This is a very quick and dirty refactor, and it's not complete. It's just to give you an idea of how it could be done):

```python
class WrapUnwrapProcessor:

    def __init__(self, cfg: Config):
        self.cfg = cfg

    def add_wrap_or_unwrap_trades_to_route(self, trade_instructions: List[TradeInstruction], route_struct: List[Dict], flashloan_struct: List) -> List[Dict]:
        """
        This method takes in trade instructions, route struct and flashloan struct and returns a new route struct with wrap/unwrap trades added.

        Args:
            trade_instructions: List[TradeInstruction], the trade instructions for the route
            route_struct: List[Dict], the route struct for the route
            flashloan_struct: List, the flashloan struct for the route

        Returns:
            List[Dict], the new route struct with wrap/unwrap trades added

        """
        self._init_balance_tracker(flashloan_struct)
        self._ensure_unique_flashloan_token()
        segmented_routes = self._segment_routes(trade_instructions, route_struct)
        return self._generate_new_route_struct(segmented_routes, route_struct, trade_instructions)

    def _init_balance_tracker(self, flashloan_struct: List):
        """
        Initialize the balance tracker based on the flashloan struct

        Args:
            flashloan_struct: List, the flashloan struct for the route
        """
        self.balance_tracker = {}
        self.flashloan_native_gas_token = False
        self.flashloan_wrapped_gas_token = False

        for flash in flashloan_struct:
            for idx, token in enumerate(flash["sourceTokens"]):
                self.balance_tracker[token] = flash["sourceAmounts"][idx]
                if token in [self.cfg.NATIVE_GAS_TOKEN_ADDRESS, self.cfg.WRAPPED_GAS_TOKEN_ADDRESS]:
                    self.flashloan_native_gas_token = True

    def _ensure_unique_flashloan_token(self):
        assert not (self.flashloan_wrapped_gas_token and self.flashloan_native_gas_token), "Cannot flashloan both wrapped & native gas tokens!"

    def _segment_routes(self, trade_instructions: List[TradeInstruction], route_struct: List[Dict]) -> Dict:
        """
        Segment the routes based on the trade instructions

        Args:
            trade_instructions: List[TradeInstruction], the trade instructions for the route
            route_struct: List[Dict], the route struct for the route

        Returns:
            Dict, the segmented routes based on the trade instructions
        """
        segmented_routes = {}
        for idx, route in enumerate(route_struct):
            real_tkn_in, real_tkn_out = trade_instructions[idx].get_real_tokens()
            pair = f"{real_tkn_in}/{real_tkn_out}"
            if pair not in segmented_routes:
                segmented_routes[pair] = {"amt_out": 0, "amt_in": 0, "token_in": real_tkn_in, "token_out": real_tkn_out, "trades": {}}

            segmented_routes[pair]["amt_out"] += trade_instructions[idx].amtout_wei
            segmented_routes[pair]["amt_in"] += trade_instructions[idx].amtin_wei
            segmented_routes[pair]["trades"][idx] = trade_instructions[idx].exchange_name
        return segmented_routes

    def _generate_new_route_struct(self, segmented_routes: Dict, route_struct: List[Dict], trade_instructions: List[TradeInstruction]) -> List[Dict]:
        """
        Generate the new route struct with wrap/unwrap trades added

        Args:
            segmented_routes: Dict, the segmented routes based on the trade instructions
            route_struct: List[Dict], the route struct for the route
            trade_instructions: List[TradeInstruction], the trade instructions for the route

        Returns:
            List[Dict], the new route struct with wrap/unwrap trades added
        """
        new_route_struct = []
        deadline = route_struct[0]["deadline"]

        for segment in segmented_routes.values():
            self._handle_wrap_or_unwrap(segment, deadline, new_route_struct)
            self._append_trades_based_on_exchange(segment, route_struct, new_route_struct)

        self._handle_final_balance(deadline, new_route_struct)
        return new_route_struct

    def _handle_wrap_or_unwrap(self, segment: Dict, deadline: int, new_route_struct: List[Dict]):
        """
        Check if wrapping or unwrapping is needed based on balance and token type, then generate wrap/unwrap structure and adjust balances

        Args:
            segment: Dict, the segment based on the trade instructions
            deadline: int, the deadline for the route
            new_route_struct: List[Dict], the new route struct with wrap/unwrap trades added
        """
        token_in = segment['token_in']
        amount_in = segment['amt_in']

        if amount_in > self.balance_tracker.get(token_in, 0):
            is_wrapping = token_in == self.cfg.WRAPPED_GAS_TOKEN_ADDRESS
            new_route_struct.append(
                {
                    "deadline": deadline,
                    "wrap": is_wrapping,
                    "source_amount": amount_in,
                }
            )
            self._adjust_balance_for_wrap_unwrap(is_wrapping, amount_in)

    def _adjust_balance_for_wrap_unwrap(self, is_wrapping: bool, amount: int):
        """
        Adjust the balance based on wrap/unwrap

        Args:
            is_wrapping: bool, whether wrapping or unwrapping
            amount: int, the amount to wrap/unwrap
        """
        token_in = self.cfg.NATIVE_GAS_TOKEN_ADDRESS if is_wrapping else self.cfg.WRAPPED_GAS_TOKEN_ADDRESS
        token_out = self.cfg.WRAPPED_GAS_TOKEN_ADDRESS if is_wrapping else self.cfg.NATIVE_GAS_TOKEN_ADDRESS
        self.balance_tracker[token_in] -= amount
        self.balance_tracker[token_out] += amount

    def _append_trades_based_on_exchange(self, segment: Dict, route_struct: List[Dict], new_route_struct: List[Dict]):
        """
        Sort and extend new_route_struct with trades based on their exchange names' presence in CARBON_V1_FORKS

        Args:
            segment: Dict, the segment based on the trade instructions
            route_struct: List[Dict], the route struct for the route
            new_route_struct: List[Dict], the new route struct with wrap/unwrap trades added
        """
        new_route_struct.extend(
            sorted(
                (route_struct[trade_idx] for trade_idx in segment['trades']),
                key=lambda x: x['exchange_name'] not in self.cfg.CARBON_V1_FORKS
            )
        )

    def _handle_final_balance(self, deadline: int, new_route_struct: List[Dict]):
        """
        Determine if wrapping or unwrapping is needed based on the balance and flashloan token type, then add the wrap or unwrap operation if needed

        Args:
            deadline: int, the deadline for the route
            new_route_struct: List[Dict], the new route struct with wrap/unwrap trades added
        """
        should_wrap = self.flashloan_wrapped_gas_token and self.balance_tracker.get(self.cfg.NATIVE_GAS_TOKEN_ADDRESS, 0) > 0
        should_unwrap = self.flashloan_native_gas_token and self.balance_tracker.get(self.cfg.WRAPPED_GAS_TOKEN_ADDRESS, 0) > 0

        if should_wrap or should_unwrap:
            new_route_struct.append({
                "deadline": deadline,
                "wrap": should_wrap or not should_unwrap,
                "source_amount": 0,
            })
```


In summary, while `add_wrap_or_unwrap_trades_to_route_v4` may perform its intended functionality, it presents a more readable, maintainable, and best-practices-aligned approach when refactored into a class. Further enhancements, especially around error handling, type annotations, and unit testing, would make the refactored class even more robust.

File: fastlane_bot/helpers/routehandler.py [hunk 2]
```diff
@@ -658,6 +753,73 @@ def _extract_flashloan_tokens(self, trade_instructions: List[TradeInstruction])
 
         return flash_tokens
 
+    def split_carbon_trades(self, trade_instructions: List[TradeInstruction]) -> List[TradeInstruction]:
```
Comment (mikewcasale) [inline fastlane_bot/helpers/routehandler.py:756]: Similar to `add_wrap_or_unwrap_trades_to_route_v4`, the `split_carbon_trades` function could also be refactored into a class following the same principles. Here is a quick and dirty refactor of the `split_carbon_trades` function into a class `CarbonTradeSplitter`:

```python
class CarbonTradeSplitter:
    NATIVE = 'native'
    WRAPPED = 'wrapped'

    def __init__(self, ConfigObj):
        self.ConfigObj = ConfigObj

    def split_carbon_trades(self, trade_instructions: List[TradeInstruction]) -> List[TradeInstruction]:
        """
        Split Carbon trades based on gas token type or exchange differences.
        """
        new_trade_list = []
        for trade in trade_instructions:
            if not self._is_carbon_trade(trade):
                new_trade_list.append(trade)
                continue

            carbon_exchanges = self._process_carbon_trades(trade)
            self._create_new_trades_from_carbon_exchanges(carbon_exchanges, trade, new_trade_list)

        return new_trade_list

    def _is_carbon_trade(self, trade: TradeInstruction) -> bool:
        """
        Check if a trade is a Carbon trade.
        """
        return trade.exchange_name in self.ConfigObj.CARBON_V1_FORKS

    def _process_carbon_trades(self, trade: TradeInstruction) -> Dict:
        """
        Process Carbon trades and organize them by exchange and token type.
        """
        carbon_exchanges = {}
        raw_txs = json.loads(trade.raw_txs)

        for _tx in raw_txs:
            curve = trade.db.get_pool(cid=str(_tx["cid"]).split('-')[0])
            exchange = curve.exchange_name
            token_type = self._get_token_type(curve, trade)

            if exchange not in carbon_exchanges:
                carbon_exchanges[exchange] = self._initialize_exchange_data()

            self._update_exchange_data(carbon_exchanges[exchange], token_type, _tx, trade)

        return carbon_exchanges

    def _get_token_type(self, curve, trade) -> str:
        """
        Determine if the trade uses native or wrapped gas tokens.
        """
        if self.ConfigObj.NATIVE_GAS_TOKEN_ADDRESS in curve.get_tokens:
            return self.NATIVE
        return self.WRAPPED

    def _initialize_exchange_data(self) -> Dict:
        """
        Initialize data structure for a new exchange.
        """
        return {
            self.NATIVE: self._new_trade_data_structure(),
            self.WRAPPED: self._new_trade_data_structure()
        }

    def _new_trade_data_structure(self) -> Dict:
        """
        Initialize a new trade data structure.
        """
        return {
            "raw_txs": [],
            "amt_in": 0,
            "amt_out": 0,
            "amtin_wei": 0,
            "amtout_wei": 0,
        }

    def _update_exchange_data(self, exchange_data: Dict, token_type: str, _tx: Dict, trade: TradeInstruction):
        """
        Update exchange data with information from a transaction.
        """
        data = exchange_data[token_type]
        data["raw_txs"].append(_tx)
        data["amt_in"] += _tx["amtin"]
        data["amt_out"] += _tx["amtout"]
        data["amtin_wei"] += _tx["_amtin_wei"]
        data["amtout_wei"] += _tx["_amtout_wei"]

        # Assuming tknin and tknout logic needs refinement based on actual requirements
        data["tknin"] = trade.tknin
        data["tknout"] = trade.tknout

    def _create_new_trades_from_carbon_exchanges(self, carbon_exchanges: Dict, original_trade: TradeInstruction, new_trade_list: List[TradeInstruction]):
        """
        Create new TradeInstruction instances from processed Carbon exchanges data.
        """
        for exchange, data in carbon_exchanges.items():
            for token_type, trade_data in data.items():
                if trade_data["raw_txs"]:
                    trade_data['cid'] = original_trade.cid
                    trade_data['db'] = original_trade.db
                    trade_data['ConfigObj'] = original_trade.ConfigObj
                    new_trade_list.append(TradeInstruction(**trade_data))
```

**Review by mikewcasale — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/tradeinstruction.py [hunk 3]
```diff
@@ -163,6 +163,30 @@ def __post_init__(self):
...
+    def get_real_tkn_out(self):
+        """
+        Returns the correct token received by the trade.
+        """
+        if self.tknout_is_native:
+            return self.ConfigObj.NATIVE_GAS_TOKEN_ADDRESS
+        elif self.tknout_is_wrapped:
+            return self.ConfigObj.WRAPPED_GAS_TOKEN_ADDRESS
+        else:
+            return self.tknout
```
Comment (barakman) [inline fastlane_bot/helpers/tradeinstruction.py:188]: Same code duplicated in two functions.
You can implement a single function which takes a token as input, then call it with `self.tknin` or `self.tknout` as needed.

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/routehandler.py [hunk 4]
```diff
@@ -440,6 +438,103 @@ def add_wrap_or_unwrap_trades_to_route(self, trade_instructions: List[TradeInstr
...
+                if exchange_name in self.ConfigObj.CARBON_V1_FORKS:
+                    new_route_struct.append(route_struct[int(trade_idx)])
+            for trade_idx in segmented_routes[segment]["trades"].keys():
+                exchange_name = trade_instructions[int(trade_idx)].exchange_name
+                if exchange_name not in self.ConfigObj.CARBON_V1_FORKS:
+                    new_route_struct.append(route_struct[int(trade_idx)])
+
+            balance_tracker[_tkn_in] -= segmented_routes[segmented_routes_in_order[idx]]["amt_in"]
+            balance_tracker[_tkn_out] += segmented_routes[segmented_routes_in_order[idx]]["amt_out"]
+            if idx == len(segmented_routes_in_order) - 1:
```
Comment (barakman) [inline fastlane_bot/helpers/routehandler.py:522]: This takes place only at the last iteration, so why not do it outside the loop?

File: fastlane_bot/helpers/routehandler.py [hunk 5]
```diff
@@ -658,6 +753,73 @@ def _extract_flashloan_tokens(self, trade_instructions: List[TradeInstruction])
...
+                new_trade_list.append(trade)
+                continue
+
+            if trade.tknin_address not in [trade.ConfigObj.NATIVE_GAS_TOKEN_ADDRESS, trade.ConfigObj.WRAPPED_GAS_TOKEN_ADDRESS] and trade.tknout_address not in [trade.ConfigObj.NATIVE_GAS_TOKEN_ADDRESS, trade.ConfigObj.WRAPPED_GAS_TOKEN_ADDRESS]:
+                new_trade_list.append(trade)
+                continue
+
+            carbon_exchanges = {}
+
+            for _tx in eval(trade.raw_txs):
```
Comment (barakman) [inline fastlane_bot/helpers/routehandler.py:777]: Can you see if there's a way to avoid using `eval` here?

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/carbon_trade_splitter.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        if token_address not in [
+            self.ConfigObj.NATIVE_GAS_TOKEN_ADDRESS,
+            self.ConfigObj.WRAPPED_GAS_TOKEN_ADDRESS,
+        ]:
+            return token_address
+        return (
+            self.ConfigObj.NATIVE_GAS_TOKEN_ADDRESS
+            if token_type == self.NATIVE
+            else self.ConfigObj.WRAPPED_GAS_TOKEN_ADDRESS
+        )
```
Comment (barakman) [inline fastlane_bot/helpers/carbon_trade_splitter.py:78]: This is a baffling implementation IMO.

The `token_type` input argument can be one of three (NATIVE or WRAPPED or NEITHER), but here you are referring to it as if it was one of two (NATIVE or not NATIVE).

This function "feels" like it was designated to handle a variety of different scenarios, but either some of these scenarios aren't valid, or the function itself has been overly-complicated.

As it stands, this function takes 2 input arguments, each of which can attain one of 3 different values:
1. Input argument `token_address`, which can be either the address of the native token or the address of the wrapped token or neither
2. Input argument `token_type`, which can be either NATIVE or WRAPPED or NEITHER.

This leads to an extremely obfuscated implementation scheme.

For example, why would this function ever receive the address of the native token or the address of the wrapped token, AND type NEITHER?

Please rewrite the function.

File: fastlane_bot/helpers/carbon_trade_splitter.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        return carbon_exchanges
+
+    def _get_token_type(self, curve) -> str:
+        """Determines if the trade uses native or wrapped gas tokens.
+
+        Args:
+            curve: the Pool object representing the curve.
+
+        Returns:
+            A string indicating if the curve contains native gas tokens, wrapped gas tokens, or neither.
```
Comment (barakman) [inline fastlane_bot/helpers/carbon_trade_splitter.py:119]: What if it contains both?

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/carbon_trade_splitter.py (hunk 7 above)
Comment (Lesigh-3100) [reply fastlane_bot/helpers/carbon_trade_splitter.py:119]: I'm not aware of any pools that contain both native + wrapped gas token. This would definitely not handle them - but I don't think we need to expect them.

File: fastlane_bot/helpers/carbon_trade_splitter.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        new_trade_list = []
+        for trade in trade_instructions:
+            if not self._is_carbon_trade(trade):
+                new_trade_list.append(trade)
+                continue
+
+            carbon_exchanges = self._process_carbon_trades(trade)
+            self._create_new_trades_from_carbon_exchanges(
+                carbon_exchanges, trade, new_trade_list
+            )
```
Comment (barakman) [inline fastlane_bot/helpers/carbon_trade_splitter.py:39]: This function (`split_carbon_trades`) is in charge of returning a list of trades.
So my suggestion is that it will be the only one populating the `new_trade_list` array which it returns.
In other words, function `_create_new_trades_from_carbon_exchanges` should return an independent list of trades, and this function should concatenate that list to its (larger) list.
This function may as well call function `_process_carbon_trades` internally (or perhaps even just merge both of these functions into a single function).

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/wrap_unwrap_processor.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+            flashloan_struct: a list containing flashloan instructions
+
+        Returns:
+            The route struct - an ordered List of Dictionaries - with the added wrap/unwrap trades.
+
+        """
+        self._init_balance_tracker(flashloan_struct)
+        self._ensure_unique_flashloan_token()
+        segmented_routes = self._segment_routes(trade_instructions, route_struct)
+        return self._generate_new_route_struct(segmented_routes, route_struct)
```
Comment (barakman) [inline fastlane_bot/helpers/wrap_unwrap_processor.py:44]: Generally speaking, it is pointless to implement a non-api (internal) function which is used only once within its module.
And there are 4 such functions here.
It hides out potential improvements within what is technically a single block of execution.
Only good reason to do so, is when these functions are large, which isn't the case here.

File: fastlane_bot/helpers/wrap_unwrap_processor.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+        """Initializes the balance tracker object.
+
+        This function initializes the balance tracker with token amounts from the flashloan struct. It also adds a flag indicating if native/wrapped gas tokens are being flashloaned.
+
+        Args:
+            flashloan_struct: a list containing flashloan instructions
+        """
+        self.balance_tracker = {}
+        self.flashloan_native_gas_token = False
+        self.flashloan_wrapped_gas_token = False
```
Comment (barakman) [inline fastlane_bot/helpers/wrap_unwrap_processor.py:56]: You are adding these variables to the class instance (self), only for the sake of internal calculations which start at this function.
They are not used anywhere outside of this module, so it doesn't make any sense to do so.
You should just pass them between functions within this module.

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/wrap_unwrap_processor.py [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+
+        for flash in flashloan_struct:
+            for idx, token in enumerate(flash["sourceTokens"]):
+                self.balance_tracker[token] = flash["sourceAmounts"][idx]
+                if token == self.cfg.NATIVE_GAS_TOKEN_ADDRESS:
+                    self.flashloan_native_gas_token = True
+                elif token == self.cfg.WRAPPED_GAS_TOKEN_ADDRESS:
+                    self.flashloan_wrapped_gas_token = True
```
Comment (barakman) [inline fastlane_bot/helpers/wrap_unwrap_processor.py:64]: There are a bunch of things to fix in this loop, and then some doubts which arise about the need for the `balance_tracker` dictionary, which seems to just map tokens to themselves.
See [this comment](https://github.com/bancorprotocol/fastlane-bot/pull/390#discussion_r1499706765) on a similar place in the code.

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/bot.py [hunk 12]
```diff
@@ -1035,12 +1040,14 @@ def _handle_trade_instructions(
...
-        if self.ConfigObj.ARB_CONTRACT_VERSION >= 10:
-            route_struct = tx_route_handler.add_wrap_or_unwrap_trades_to_route(
-                trade_instructions=calculated_trade_instructions,
-                route_struct=route_struct,
-                flashloan_struct=flashloan_struct,
-            )
+
+        route_struct_processed = WrapUnwrapProcessor(cfg=self.ConfigObj).add_wrap_or_unwrap_trades_to_route(trade_instructions=split_calculated_trade_instructions, route_struct=route_struct, flashloan_struct=flashloan_struct)
+
+        route_struct = tx_route_handler.add_wrap_or_unwrap_trades_to_route_v4(
```
[the remark this answers is not in the document]
Comment (Lesigh-3100) [reply fastlane_bot/bot.py:1046]: This was a leftover - thank you - removing.

File: fastlane_bot/helpers/wrap_unwrap_processor.py [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+
+    def __init__(self, cfg: Any):
+        self.cfg = cfg
+
+    def add_wrap_or_unwrap_trades_to_route(
+        self,
+        trade_instructions: List[TradeInstruction],
+        route_struct: List[Dict],
+        flashloan_struct: List,
+    ) -> List[Dict]:
```
Comment (barakman) [inline fastlane_bot/helpers/wrap_unwrap_processor.py:27]: In file routerhandler.py, [there is a function](https://github.com/bancorprotocol/fastlane-bot/blob/381-add-ability-to-split-native-vs-wrapped-carbon-orders/fastlane_bot/helpers/routehandler.py#L354-L356) with the same name, which takes the same input arguments in the same order, and returns the same output type.
The implementation of that function looks suspiciously similar to the implementation of this function.
Consider implementing the necessary functionality in a single place in the code.

**Review by barakman — CHANGES_REQUESTED** [review]

File: resources/NBTest/NBTest_062_TestRouteHandler.py [hunk 14]
```diff
@@ -424,3 +426,439 @@
...
+        _balance[trade['sourceToken']] -= trade['sourceAmount']
+        _balance[trade['targetToken']] += trade['minTargetAmount']
+        check_no_negative(_balance)
+
+
+for idx, (_route, _balance) in enumerate([(wrap_unwrap_0_processor, balances_0_processor), (wrap_unwrap_1_processor, balances_1_processor), (wrap_unwrap_2_processor, balances_2_processor), (wrap_unwrap_3_processor, balances_3_processor)]):
+    #print(f"starting loop {idx}")
+    loop_balances(_route, _balance)
+
+## For Debugging
```
Comment (mikewcasale) [inline resources/NBTest/NBTest_062_TestRouteHandler.py:830]: All commented out code sections should be removed before we merge changes.

File: fastlane_bot/helpers/wrap_unwrap_processor.py [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+                deadline=deadline,
+                customAddress=self.cfg.WRAPPED_GAS_TOKEN_ADDRESS
+                if wrap
+                else self.cfg.NATIVE_GAS_TOKEN_ADDRESS,
+                customInt=0,
+                customData="0x",
+            )
+        )
+
+    class FlashloanTokenException(Exception):
```
Comment (mikewcasale) [inline fastlane_bot/helpers/wrap_unwrap_processor.py:341]: If this is a new exception class, we should get in the habit of defining these in a dedicated `exceptions.py` module. If it's not newly defined in this PR you can disregard.

**Review by mikewcasale — DISMISSED** [review]

File: resources/NBTest/NBTest_062_TestRouteHandler.py (hunk 14 above)
Comment (Lesigh-3100) [reply resources/NBTest/NBTest_062_TestRouteHandler.py]: Removed commented code

File: resources/NBTest/NBTest_070_TestCarbonTradeSplitter.py
[the remark this answers is not in the document]
Comment (Lesigh-3100) [reply resources/NBTest/NBTest_070_TestCarbonTradeSplitter.py:1]: Removed direct function calls.

File: fastlane_bot/helpers/wrap_unwrap_processor.py (hunk 15 above)
Comment (Lesigh-3100) [reply fastlane_bot/helpers/wrap_unwrap_processor.py]: Moved FlashloanTokenException -> exceptions.py

File: fastlane_bot/helpers/wrap_unwrap_processor.py (hunk 13 above)
Comment (Lesigh-3100) [reply fastlane_bot/helpers/wrap_unwrap_processor.py:28]: Removed the now-unused code from Routehandler.

File: fastlane_bot/helpers/wrap_unwrap_processor.py (hunk 9 above)
Comment (Lesigh-3100) [reply fastlane_bot/helpers/wrap_unwrap_processor.py:45]: Splitting it out into functions made it much easier to test the individual processes - previously this was one single giant function.

**Review by barakman — CHANGES_REQUESTED** [review]

**Review by barakman — DISMISSED** [review]

**Review by zavelevsky — APPROVED** [review]
LGTM!

**Review by barakman — APPROVED** [review]
